diff -ru libressl-3.5.2/crypto/asn1/x_crl.c libressl/crypto/asn1/x_crl.c
--- libressl-3.5.2/crypto/asn1/x_crl.c	2022-03-16 02:37:03.000000000 +1000
+++ libressl/crypto/asn1/x_crl.c	2022-06-01 10:30:47.075625844 +1000
@@ -716,6 +716,12 @@
 	return crl->crl->lastUpdate;
 }
 
+const X509_ALGOR *
+X509_CRL_get0_tbs_sigalg(const X509_CRL *crl)
+{
+	return (crl->crl->sig_alg);
+}
+
 ASN1_TIME *
 X509_CRL_get_lastUpdate(X509_CRL *crl)
 {
diff -ru libressl-3.5.2/include/openssl/x509.h libressl/include/openssl/x509.h
--- libressl-3.5.2/include/openssl/x509.h	2022-04-21 03:52:14.000000000 +1000
+++ libressl/include/openssl/x509.h	2022-06-01 10:36:54.728086352 +1000
@@ -412,6 +412,7 @@
 STACK_OF(X509_REVOKED) *X509_CRL_get_REVOKED(X509_CRL *crl);
 void X509_CRL_get0_signature(const X509_CRL *crl, const ASN1_BIT_STRING **psig,
     const X509_ALGOR **palg);
+const X509_ALGOR *X509_CRL_get0_tbs_sigalg(const X509_CRL *crl);
 
 int X509_REQ_get_signature_nid(const X509_REQ *req);
 
